### DISCUSSION TITLE
ostree-prepare-root generates a kernel panic.

### DISCUSSION THREAD

**Discussion by rmsaad** [ostreedev/ostree#3251] [Q&A] [answered]
Hello, 
I have been trying develop an understanding for how ostree works by playing around with qemu + uboot. I have created a filesystem which contains the ostree deployment and created a qemu image from it. 
```
    ostree admin --sysroot=${SYSROOT_DIR} init-fs ${SYSROOT_DIR}; 
    ostree admin --sysroot=${SYSROOT_DIR} os-init ${OS_NAME}; 
    mkdir -p ${SYSROOT_DIR}/boot/loader.0; 
    ln -s loader.0 ${SYSROOT_DIR}/boot/loader; 
    touch ${SYSROOT_DIR}/boot/loader/uEnv.txt; 
    ostree init --repo=${SYSROOT_DIR}/${REPO_DIR}; 
    ostree remote add --repo=${SYSROOT_DIR}/${REPO_DIR} --gpg-import=${GPG_KEY} origin ${CLOUD_LOC}/${CLOUD_IMG} ${OSTREE_BRANCH}; 
    ostree pull origin --repo=${SYSROOT_DIR}/${REPO_DIR} ${OSTREE_BRANCH}; 
    ostree admin --sysroot=${SYSROOT_DIR} deploy --os=${OS_NAME} origin:${OSTREE_BRANCH};
    mkimage -T script -C none -n "Boot script" -d /boot.txt ${SYSROOT_DIR}/boot/boot.scr;
```

boot script:
```
if test -e ide 0:1 /boot/loader/uEnv.txt; then
    echo Booting OSTree ...;    
    load ide 0:1 1000000 /boot/loader/uEnv.txt;
    env import -t 1000000 ${filesize};
    echo "Loaded uEnv.txt with filesize: ${filesize}";
    echo "Kernel Image: ${kernel_image}";
    
    ext2load ide 0:1 1000000 ${kernel_image}
    if test $? -ne 0; then
        echo "Failed to load kernel image: ${kernel_image}";
    fi;

    setenv bootargs root=/dev/sda1 console=ttyS0 rw rootwait tsc=unstable ${bootargs};
    echo "Boot arguments: ${bootargs}";
    zboot
fi;
```

When I try to launch the image using qemu, I get the following kernel panic. I feel like the fact that I am not using an initramfs might be related, the ostree-prepare-root documentation makes it sound like it is necessary "This tool is designed to run in an initramfs", but other ostree documentation makes it sound like an initramfs is optional.

```
user@hostname:~/ostree$ qemu-system-x86_64 -bios u-boot/u-boot.rom -drive file=rootfs.img,format=raw -m 1G -nographic
...
[    2.255017] Run /ostree/boot.1/debian/f2fb8ef17b1d73c07991ba30a4af97e1417caa0176e2fc8e475c4909d9b58168/0/usr/lib/ostree/ostree-prepare-root as init process
[    2.255471]   with arguments:
[    2.255554]     /ostree/boot.1/debian/f2fb8ef17b1d73c07991ba30a4af97e1417caa0176e2fc8e475c4909d9b58168/0/usr/lib/ostree/ostree-prepare-root
[    2.255837]   with environment:
[    2.255922]     HOME=/
[    2.256008]     TERM=linux
[    2.256084]     ostree=/ostree/boot.1/debian/f2fb8ef17b1d73c07991ba30a4af97e1417caa0176e2fc8e475c4909d9b58168/0
[    2.268547] Kernel panic - not syncing: Requested init /ostree/boot.1/debian/f2fb8ef17b1d73c07991ba30a4af97e1417caa0176e2fc8e475c4909d9b58168/0/usr/lib/ostree/ostree-prepare-root failed (error -2).
[    2.269212] CPU: 0 PID: 1 Comm: swapper/0 Tainted: G          I       5.10.201 #1
[    2.269354] Hardware name: emulation qemu-x86/qemu-x86, BIOS 2024.01-rc2-00034-ge17d174773 01/01/2024
[    2.269354] Call Trace:
[    2.269354]  dump_stack+0x57/0x6e
[    2.269354]  ? rest_init+0x40/0xa4
[    2.269354]  panic+0x109/0x2ca
[    2.269354]  ? rest_init+0xa4/0xa4
[    2.269354]  kernel_init+0xab/0x10c
[    2.269354]  ret_from_fork+0x22/0x30
[    2.269354] Kernel Offset: 0x20400000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[    2.269354] ---[ end Kernel panic - not syncing: Requested init /ostree/boot.1/debian/f2fb8ef17b1d73c07991ba30a4af97e1417caa0176e2fc8e475c4909d9b58168/0/usr/lib/ostree/ostree-prepare-root failed (error -2). ]---
```

What I also find weird is I am able to install ostree from inside a normal debian image running in qemu and deploy, but when I start  removing certain directories such as sysroot/usr and sysroot/proc and reboot I will start to get kernel panics again (my understanding is only /boot and /ostree are necessary on sysroot).

A nudge in the right direction would be greatly appreciated.

**Comment by ericcurtin** [comment]
ostree-prepare-root as an init process isn't gonna work, it's just not designed to be an init system.

What do you plan on using for init? systemd or something else?

Like even if we added the code to turn it into an init process, it's just beyond the scope of what is planned for OSTree.

What you probably want to do is the equivalent of this in your init system:

https://github.com/ostreedev/ostree/blob/main/src/boot/ostree-prepare-root.service

**Reply by rmsaad** [reply]
Yes I am using systemd. I will take a look into that service.

I am confused because ostree generated uboot bootargs, I am guessing from here:
https://github.com/ostreedev/ostree/blob/64a09da0eb5ca4dade83125a2ebc47b52c60c5a5/src/libostree/ostree-bootloader-uboot.c#L147

```
bootargs=init=/ostree/boot.1/debian/f2fb8ef17b1d73c07991ba30a4af97e1417caa0176e2fc8e475c4909d9b58168/0/usr/lib/ostree/ostree-prepare-root ostree=/ostree/boot.1/debian/f2fb8ef17b1d73c07991ba30a4af97e1417caa0176e2fc8e475c4909d9b58168/0
```

I made the assumption that it was sufficient to boot with them from a uboot script.

**Reply by ericcurtin** [reply]
Btw if you are doing this to learn about OSTree and just want as OS to quickly start hacking on in qemu, highly recommend Fedora IoT, Silverblue or Kinoite

**Comment by rmsaad** [comment, marked as answer]
I managed to fix my problem. ostree-prepare-root was looking for shared libraries which are only available in the deployment itself. Compiling ostree with --with-static-compiler="gcc" fixed the issue.
